### PR TITLE
Update lslboost

### PR DESCRIPTION
**Pull request by tstenner**
As mentioned in #227, this updates lslboost to 1.65.1. It's about a third of the size of the current lslboost, but has a latency problem (at least on windows).

### CODE REVIEW

**Comment by dmedine** [comment]
By the way, what is the nature of this latency problem on Windows? I presume this isn't a problem on other platforms?

**Comment by tstenner** [comment]
Oversimplified: to reduce the power usage, Windows batches sleep requests so the system can wake up, do the work and then get back to a lower power state. To do that, `sleep()` takes the time to sleep more as a general guideline and "corrects" all sleep requests below 27(?)ms. Other platforms assume that you know what you're doing and try to wake up your thread after the desired time.

**Comment by dmedine** [comment]
There must be a winapi call to change this behavior. I will poke around. 
I wonder why this is only the case in more modern boost deployments? My 
guess would be that the libraries are calling some newer sleep() 
function that comes in Windows 10 or something like that.

**Comment by tstenner** [comment]
Currently, it's only in Boost (see [here](https://stackoverflow.com/a/30414877)), so we could patch it out for the moment.

**Comment by dmedine** [comment]
Ah, so it was a deliberate choice on the Boost end. Then there will be 
no simple winapi flag to flip. This should definitely be patched, 
though, since low-latency in LSL takes priority over battery life. I 
will take care of this. In the meantime can merge this pull before it 
get's done, though. In the meantime, windows folks can be content with 
old boost or higher latencies.

**Comment by tstenner** [comment]
I'll rebase it first and run it through the CI systems, just to be on the safe side.

File: LSL/liblsl/external/boost/asio/deadline_timer.hpp [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#include <lslboost/asio/deadline_timer.hpp>
```
Comment (tstenner) [inline LSL/liblsl/external/boost/asio/deadline_timer.hpp]: Some files include all of asio/thread but need only a specific header. This commit adds those headers.

**Comment by dmedine** [comment]
Maybe I misunderstood this. I thought you were working on tools so that 
a developer can choose any version of boost to be lsl-erized and used as 
lsl_boost. Do I have this wrong? Looks like you are hardcoding boost 
1.64 (or whatever newer version), which is very good, of course, but 
ultimately, the flexibility to choose is better still.

**Comment by dmedine** [comment]
Ahh. After reviewing issue 227 (https://github.com/sccn/labstreaminglayer/issues/227) I see what is happening now. This is exactly as you and Matthew had discussed. 

I would still like to see an 'any boost' solution, but that is probably a different thread.

**Comment by tstenner** [comment]
Most users who require a newer version (e.g. on platforms with no asio support in the old lslboost) can just use the included newer lslboost, and for them it's just one build setting.

**Comment by tstenner** [comment]
The CI still has some complains with the rebased commits, I'll look into it tomorrow

**Comment by tstenner** [comment]
The CI builds pass: [Travis](https://travis-ci.org/tstenner/labstreaminglayer/builds/298962913), [Appveyor](https://ci.appveyor.com/project/tstenner/labstreaminglayer/build/1.40.53). The old lslboost is still the default, so nothing should change for regular users.

**Comment by dmedine** [comment]
Sorry I am unable to find update_lslboost.bat/sh. Where is this file?

**Comment by tstenner** [comment]
I added them in 521bedc, but I tested the Unix version more. Changing the minimum wait time in Boost (442ab45) also reduces the average latency in the benchmark (Bounce.cpp in c37226d) from 15ms to 1.8ms, so this could be all we need.

**Comment by dmedine** [comment]
Ok, I think I am finally up to speed on all the work (big ups, btw!!) you've done here. 

Have yout tested your boost::thread patch on Windows? If not, I can do it today. I have to prepare a 64-bit version of an app for someone anyway, so I can measure the sleep time while I'm at it. Even if it does work, I'm inclined to go another route. Let me know your thoughts, but I think what is safest is to change the mangler some.  What I would do is to add the behavior that if a developer wishes to use a non-old version of boost, she can call the bat file (or python script---this is much easier make with python) that will mangle a specified boost installation, but (if it is >=1.58) retain the old version of boost::thread. This, of course, means keeping the old version of boost::thread around, which is admittedly ugly.

Let me know your thoughts.

**Comment by dmedine** [comment]
Also, I'm not quite sure what to do about the missing headers in bcp. Maintaining a master list and checking against it might be necessary. Doing it by hand is a drag, especially if people are needing to re-compile against certain boost versions to suit their needs.

**Comment by tstenner** [comment]
I tested the sleep duration (with the patch) on Windows Server 2012 and it didn't notice any problems so the 4 changed lines in 442ab45 should be all that's needed.

The process to add a custom boost version looks like this:

1. Download Boost and extract it somewhere
2. Extract a subset with bcp* to a subdirectory of `LSL/liblsl`
3. Copy the build files (`CMakeLists.txt`) from `lslboost` to this new folder and change them**
4. Change `LSL_LSLBOOST_PATH` to the name of the subfolder
5. If necessary, change Boost.Thread***

There are three ways to improve this:  
`*` On Windows, the included boost files have to be enumerated in `used_headers.h` so bcp knows which boost headers are included, but it can also accept a list of all lsl source files.
On Unix, the names of all LSL source files are appended to the command line (`find src -regex ".+\.[ch]p*"`) automatically. It might be possible to do something like this in batch files or to automatically or generate the file: (`git grep -h '^#include <boost/' -- src/ > lslboost/used_headers.h`, i.e. "find all lines in `src` that start with `#include <boost/` and put them in `lslboost/used_headers.h`).

`**`: the `lslboost_headers` variable isn't strictly needed, it just enables rebuilds when a boost header changes. The remaining lists with cpp-files are a bit of a hassle, but I don't see any way to automate this.

`***`: The patch is just 4 lines, but if it gets more, we could supply a list of patches / commits to cherry-pick we recommend.

I wouldn't change much, just add documentation. This already covers the "official" lslboost version and a current boost version. I'm not sure which users know they need a special boost version but aren't able to follow these instructions, so I wouldn't put too much work into it.

**Comment by dmedine** [comment]
Ok. I will test this on Windows 10 today. If the patch is rock solid (as it appears to be), then I'm happy to keep this solution.

As for the three improvements:
* I will repeat my advocacy for a Python solution to this as it eliminates need to maintain different scripts for different os's and is much more agile at these kinds of tasks anyway. If agreed, then I will take care of this.

** requires more thought

*** git already maintains a commit history, can we just assume that anyone that needs to deal with this stuff can dig into that? Certainly this should be mentioned in any forthcoming documentation (which I'm glad to help out with as well).

**Comment by tstenner** [comment]
`*`: To be honest, with a good documentation the scripts aren't needed so I'd remove them altogether or require bash on Windows (either msys or the the Linux subsystem for Windows). Python is great, but kind of verbose and un-pythonic for shell scripting.

`***`: The documentation could include a list of commits with a short explanation / link to the bug report (e.g. `442ab45`: see #227). It would be nice to only see the relevant commits, as even `git blame` would list every commit updating the entire official lslboost.

**Comment by tstenner** [comment]
FYI, I rebased this so you'll need to force pull the branch.

**Comment by dmedine** [comment]
Cool. I'd like to do some more tests before merging. Mainly I want to try the sleep duration test on Windows 10. I promised to do this last week, but became immersed with cmake.

**Comment by tstenner** [comment]
In the configuration code, 2 functions from Boost.Filesystem were used, replacing them with a helper function removes the dependency on about 7.5K lines of code and almost half of the platform specific Boost objects

**Comment by cboulay** [comment]
On MacOS 10.13.2 (High Sierra)

| boost folder | M (ms) | SD (ms) |
| |:- |:- |
| lslboost | 1.4664 | 0.369024 |
| external | 1.44421 | 0.332669 |

**Comment by tstenner** [comment]
I've tested the code with the old Boost branch (which is still the default) and even VS2008 is happy with it, so I'd merge the PR at the end of next week if there are no objections.

**Comment by dmedine** [comment]
Jawohl.

**Pull request merged**